crates/task-impls/src/response.rs

```diff
@@ -0,0 +1,179 @@
+use std::{collections::BTreeMap, sync::Arc};
+
+use async_broadcast::Receiver;
+use async_compatibility_layer::art::async_spawn;
+use async_lock::RwLock;
+use bincode::Options;
+use either::Either::Right;
+use futures::{channel::mpsc, FutureExt, StreamExt};
+use hotshot_constants::VERSION_0_1;
+use hotshot_task::dependency::{Dependency, EventDependency};
+use hotshot_types::{
+    consensus::Consensus,
+    data::VidDisperse,
+    message::{
+        CommitteeConsensusMessage, DataMessage, Message, MessageKind, Proposal, SequencingMessage,
+    },
+    traits::{
+        election::Membership,
+        network::{DataRequest, RequestKind, ResponseChannel, ResponseMessage},
+        node_implementation::NodeType,
+        signature_key::SignatureKey,
+    },
+};
+use hotshot_utils::bincode::bincode_opts;
+use sha2::{Digest, Sha256};
+
+use crate::events::HotShotEvent;
+
+/// Type alias for consensus state wrapped in a lock.
+type LockedConsensusState<TYPES> = Arc<RwLock<Consensus<TYPES>>>;
+
+/// Type alias for the channel that we receive requests from the network on.
+type ReqestReceiver<TYPES> = mpsc::Receiver<(Message<TYPES>, ResponseChannel<Message<TYPES>>)>;
+
+/// Task state for the Network Request Task. The task is responsible for handling
+/// requests sent to this node by the network.  It will validate the sender,
+/// parse the request, and try to find the data request in the consensus stores.
+pub struct NetworkRequestState<TYPES: NodeType> {
+    /// Locked consensus state
+    consensus: LockedConsensusState<TYPES>,
```

Comment:
Does this allow us to basically dispatch to consensus when we receive certain responses? How come we don't just broadcast in this case?

Reply:
It's the opposite, it allows us to read from the consensus state and respond to whoever is requesting the data.  The requests here are coming from external parties over the network layer.